Automattic/wp-calypso
# isSiteChecklistComplete: Only blog owners need personal tasks to be completed

#### Changes proposed in this Pull Request

The `isSiteChecklistComplete` redux selector will now also check to see if you're the blog's owner.
* If you are the blog's owner, there is no change and the selector acts just the same as before.
* If you are not the blog's owner, the selector ignores any tasks with `.taskType == "user"`.  (In practice, this means it only needs `.taskType == "blog"` tasks to be completed.)

### Issue

The issue being solved is users sometimes seeing a incomplete "setup"
bar adjacent to "My Home" in the left-side bar, but the main area
doesn't let them know what tasks they still need to do.  This only
happens for blogs with multiple admins.

![2020-07-29_15-02](https://user-images.githubusercontent.com/937354/88847621-e5e82d80-d1ac-11ea-93b2-9e8b4c393597.png)


#### Testing instructions

- Create a new blog on your main WordPress.com account.
- Go to calypso and finish all of the setup steps.
- Notice that the setup bar in the left sidebar has disappeared, as well
  as the task list in the main area.
- In another browser or device, create a new WordPress.com account, or
  have one handy that has never completed the full checklist.  It's okay
  to verify your email, but do not download the mobile app for the
  second account.
- Back on the first account, go to Manage -> People and invite your
  second account to be an admin of this blog.
- On the second account, check your email and accept the invitation.
- On the second account, visit the main calypso page for this blog.
- Notice the main area is missing the task list, but in the left
  sidebar, adjacent to "My Home", you see an incomplete Setup bar.

### Why this is happening:

The task the second account is missing is a "personal" task.  For
me, the task list looked like the following:

```
[
  {
    "id": "front_page_updated",
    "isCompleted": true,
    "taskType": "blog"
  },
  {
    "id": "site_launched",
    "isCompleted": true,
    "taskType": "blog"
  },
  {
    "id": "email_verified",
    "isCompleted": true,
    "taskType": "user"
  },
  {
    "id": "mobile_app_installed",
    "isCompleted": false,
    "taskType": "user"
  }
]
```

However, the "task checklist" does not appear on the blog.  That's
because it's a per-blog setting managed by wpcom.  When the site owner
finishes the checklist, that blog gets marked as never showing the task
checklist layout again.  This is done server-side, here: r207597-wpcom ,
and also mentioned in issue #42263.

The problem is a mismatch between the sidebar "setup bar" and the main
area "task checklist".  The sidebar "setup bar" is different per-user,
and the main area "task checklist" is hidden per-blog.  When the "setup
bar" appears but the "task checklist" is hidden, the user doesn't know
what task they need to complete.

There are different ways to fix this, so this should be reviewed
carefully.  Please provide input.

#### Sidebar + Main Coordination (Not attempted)

Coordinating the left "setup bar" and the main area "task checklist"
might be possible, but I found it complex and daunting.  What appears in
the main area is a coordination between calypso and wpcom.  Some action
creators an a data layer lie here:
client/state/data-layer/wpcom/sites/home/layout/index.js

Ultimately, what's shown in the main area comes from an XHR to
https://public-api.wordpress.com/wpcom/v2/sites/999999999/home/layout

Additionally, the current code seems to work well in our common
use case (one user, one blog), so it seems to me a good idea to leave it
alone.

#### Task Filtering (This commit)

This fixes the most common case.  A site owner starts a new blog, fills
out the blog checklist, and invites a second admin.  The second admin
goes to the blog and sees a left "setup bar", but they don't know what
they need to do.

This fix makes it so if non-owners are looking at the main page, only
the blog setup tasks are considered when deciding if the "setup bar"
should be shown.

#### More information on left sidebar (Not attempted)

Since the main issue is the left sidebar mentioning tasks are not
finished without specifying what those tasks are, another possible
approach would be to have a way for the sidebar to display those tasks
if possible.  But this would duplicate information in the common case,
unless we coupled it to the current layout displayed so it could check,
which doesn't seem good imo.

#### Others?

Any other approaches would be welcomed.

114-gh-dotcom-manage
